### DISCUSSION TITLE
Saving data output to server in csv format

### DISCUSSION THREAD

**Discussion by Altr37** [jspsych/jsPsych#1061] [Q&A] [answered]
Hi,

Is there a way to save subject's results to server in csv format?

Thank you!

**Comment by jodeleeuw** [comment]
Hi @Altr37 

What kind of server are you using?

**Comment by Quinlanptq** [comment, marked as answer]
Hi I sorted this out a while back.

when my timeline expires in my on_finish routine I have this

        var pq_list = [];
		for (var ix =0; ix<pq_data.length; ix++){
            var pq_obj = pq_data[ix];
            if (pq_obj['Got_cha'] != undefined){
                
                var pq_dict ={
                    Exp_Name: pq_obj["ExpName"],
                    Part_Id :pq_obj["userId"],
                    Gender : pq_obj["Gender"],
                    Birth: pq_obj["BirthId"],
                    Block: pq_obj["Block"],
                    TBlock: pq_obj["TBlock"],
                    Cond: pq_obj["Trial_type"],
                    Correct: pq_obj["Correct"],
                    key_press: pq_obj["key_press"],
                    rt: pq_obj["rt"],

                }
                pq_list.push(pq_dict);
            }
   
		}
		    
        var pq_csv = convertArrayOfObjectsToCSV({data: pq_list});

and the convert function is this 

function convertArrayOfObjectsToCSV(args) {
    var result, ctr, keys, columnDelimiter, lineDelimiter, data;

    data = args.data || null;
    if (data == null || !data.length) {
        return null;
    }

    columnDelimiter = args.columnDelimiter || ',';
    lineDelimiter = args.lineDelimiter || '\n';

    keys = Object.keys(data[0]);

    result = '';
    result += keys.join(columnDelimiter);
    result += lineDelimiter;

    data.forEach(function(item) {
        ctr = 0;
        keys.forEach(function(key) {
            if (ctr > 0) result += columnDelimiter;

            result += item[key];
            ctr++;
        });
        result += lineDelimiter;
    });

    return result;
}

I then call
save_data_nc(out_file,pq_csv);

which looks a bit like this

function save_data_nc(name,data){
    var xhr = new XMLHttpRequest();
    xhr.open('POST', 'https://to your php file on your server '); 
    xhr.setRequestHeader('Content-Type', 'application/json');
    xhr.send(JSON.stringify({filename: name, filedata: data}));
}

Best
Philip.

**Reply by jodeleeuw** [reply]
FYI there is a built-in method that converts the internal data structure to a CSV string:

```js
jsPsych.data.get().csv()
```

**Reply by Altr37** [reply]
Thank you! I'll give it a try

**Comment by Quinlanptq** [comment]
Sorry yes I did look into that but I wanted to customise what data I wanted saving to the file hence my rather cumbersome code
